kufar/books/main.go

```diff
@@ -95,27 +133,119 @@ func New() (*Database, error) {
 	return &Database{
 		db: db,
 	}, nil
-
 }
 
 //Close closes the conn
 func (d Database) Close() error {
 	return d.db.Close()
 }
 
+//SignUpUser ...
+func (d Database) SignUpUser(w http.ResponseWriter, r *http.Request) {
+	credReq := CredentialsRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&credReq); err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError)+": error unmarshalling json", http.StatusInternalServerError)
+		return
+	}
+
+	pword, err := bcrypt.GenerateFromPassword([]byte(credReq.Password), cost)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError)+": could not add password", http.StatusInternalServerError)
+		return
+	}
+
+	cred := Credentials{Username: credReq.Username}
+	if _, err = d.db.Exec("insert into credentials(username, password) values($1, $2)", cred.Username, string(pword)); err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError)+": could not log in user. Username already exists", http.StatusInternalServerError)
+		return
+	}
+
+	w.Write([]byte(http.StatusText(http.StatusOK)))
+}
+
+//SignInUser ...
+func (d Database) SignInUser(w http.ResponseWriter, r *http.Request) {
+	credReq := CredentialsRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&credReq); err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError)+": error unmarshalling json", http.StatusInternalServerError)
+		return
+	}
+
+	row := d.db.QueryRow("select uuid, password from credentials where username = $1", credReq.Username)
+
+	cred := Credentials{}
+	err := row.Scan(&cred.UUID, &cred.Password)
+	switch {
+	case err == sql.ErrNoRows:
+		http.NotFound(w, r)
+		return
+	case err != nil:
+		http.Error(w, http.StatusText(http.StatusInternalServerError)+": could not scan db", http.StatusInternalServerError)
+		return
+	}
+
+	session, err := store.Get(r, "my-cookie")
```

Comment:
Did you test this behavior? I think in that context there will be always nil session..

Reply:
tested it. `session == nil` was always false

Reply:
and ran the endpoint multiple times, and had no issue with the session

Reply:
👍